### PR TITLE
Add new `pr-pull` command

### PR DESCRIPTION
**Pull request by jonchang**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

This pull request adds a new development command, `brew pr-pull`, which is essentially the local version of what [`brew pr-publish`](https://github.com/Homebrew/brew/pull/7202) does. Or from another perspective, it's the GitHub Actions replacement for `brew pull --bottle`.

When a maintainer runs `brew pr-pull PR_NUMBER`, this command will:

* Use the GitHub API to obtain the artifact download URL
* Unpack the artifact in a temporary directory
* Upload the bottles to Bintray
* Publish the bottles to Bintray
* Apply the pull request patches to the tap directory
* Apply the bottle commits to the tap directory

The maintainer will then need to e.g., `cd $(brew --repo homebrew/core) && git push`.

https://github.com/Homebrew/homebrew-core/pull/52305 was merged via this command.

### CODE REVIEW

**Comment by issyl0** [comment]
I'm still old-fashioned and use `brew pull --bottle` because I haven't had enough headspace to investigate `brew pr-publish` yet.

For the confused, when should we use `brew pr-publish`, and when should we use `brew pr-pull`?

**Comment by jonchang** [comment]
> when should we use brew pr-publish, and when should we use brew pr-pull?

If this gets merged, “publish” should be when you like all the commits on the head of a pull request and want to just merge it directly with bottles. On the other hand “pull” should be used when you need to modify commits or make other changes but don’t want to push to the pull request and have CI run again. 

In the future I’ll add an auto squash option to “publish” that will hopefully address most use cases of needing to modify commits when merging bottles.

**Comment by Bo98** [comment]
Would it be worth including the "do you need to pull inside master?" check from `brew pull`? That check may have saved me a couple times.

Would accepting a `--bintray-org` and passing it to test-bot be useful for anyone? I've never used it so I just wanted to raise the point.

**Comment by MikeMcQuaid** [comment]
> Would it be worth including the "do you need to pull inside master?" check from `brew pull`? That check may have saved me a couple times.

👍.

In general it may be worth reading through the (somewhat painful) `brew pull` source code for all the other "gotchas" there.

File: Library/Homebrew/dev-cmd/pr-pull.rb [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+      switch "--workflow=",
+             description: "Specify the workflow name (default: tests.yml)."
+      switch :verbose
+      switch :debug
+      min_named 1
+    end
+  end
+
+  def signoff!(pr, path: ".", dry_run: false)
+    message = `git -C #{path} log -1 --pretty=%B`
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/pr-pull.rb]: Can we generally use `Utils.popen` rather than backticks.

File: Library/Homebrew/dev-cmd/pr-pull.rb [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+      merge_base = `git -C #{path} merge-base HEAD FETCH_HEAD`.strip
+      safe_system "git", "-C", path, "cherry-pick", "#{merge_base}..FETCH_HEAD"
+    end
+  end
+
+  def pr_pull
+    pr_pull_args.parse
+
+    if !ENV["HOMEBREW_BINTRAY_USER"] || !ENV["HOMEBREW_BINTRAY_KEY"]
+      odie "Missing HOMEBREW_BINTRAY_USER or HOMEBREW_BINTRAY_KEY variables!" unless args.dry_run? || args.no_upload?
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/pr-pull.rb]:
```suggestion
      odie "Missing HOMEBREW_BINTRAY_USER or HOMEBREW_BINTRAY_KEY variables!" if !args.dry_run? && !args.no_upload?
```

assuming this is correct. `unless ... ||` is a bit hard to follow.

File: Library/Homebrew/dev-cmd/pr-pull.rb [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+      safe_system "git", "-C", path, "fetch", "--quiet", "--force", "origin", "+refs/pull/#{pr}/head"
+      merge_base = `git -C #{path} merge-base HEAD FETCH_HEAD`.strip
+      safe_system "git", "-C", path, "cherry-pick", "#{merge_base}..FETCH_HEAD"
+    end
+  end
+
+  def pr_pull
+    pr_pull_args.parse
+
+    if !ENV["HOMEBREW_BINTRAY_USER"] || !ENV["HOMEBREW_BINTRAY_KEY"]
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/pr-pull.rb]:
```suggestion
    if ENV["HOMEBREW_BINTRAY_USER"].blank? || ENV["HOMEBREW_BINTRAY_KEY"].blank?
```

File: Library/Homebrew/dev-cmd/pr-pull.rb [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+      odie "Missing HOMEBREW_BINTRAY_USER or HOMEBREW_BINTRAY_KEY variables!" unless args.dry_run? || args.no_upload?
+    end
+
+    workflow_name = args.workflow || "tests.yml"
+    test_bot_cmd = %W[#{HOMEBREW_PREFIX}/bin/brew test-bot --ci-upload]
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/pr-pull.rb]: Not super happy about a Homebrew/brew command that calls `brew test-bot`. `test-bot` is a bit of a mess (even moreso than `brew pull`). Would be nice to instead perhaps move the upload code from `test-bot` into Homebrew/brew and have it call back. In a perfect world: it'd be a separate command that `brew test-bot` can call (but this command calls the methods rather than the command) to be able to have the workflow be a bit more modular if stuff goes wrong.

File: Library/Homebrew/utils/github.rb [hunk 5]
```diff
@@ -438,6 +440,38 @@ def dispatch_event(user, repo, event, **payload)
...
+    pr_payload = open_api("#{base_url}/pulls/#{pr}")
+    pr_sha = pr_payload["head"]["sha"]
+    pr_branch = pr_payload["head"]["ref"]
+
+    workflow = open_api("#{base_url}/actions/workflows/#{workflow_id}/runs?branch=#{pr_branch}&status=completed")
+    workflow_run = workflow["workflow_runs"].select do |run|
+      run["head_sha"] == pr_sha
+    end
+
+    artifacts = open_api(workflow_run.first["artifacts_url"])
```
Comment (MikeMcQuaid) [inline Library/Homebrew/utils/github.rb]: Needs a check for `.first` being `nil` or `workflow_run` being empty.

File: Library/Homebrew/utils/github.rb [hunk 6]
```diff
@@ -438,6 +440,38 @@ def dispatch_event(user, repo, event, **payload)
...
+    onoe "More than one artifact available for PR ##{pr}" if artifacts["total_count"] > 1
```
Comment (MikeMcQuaid) [inline Library/Homebrew/utils/github.rb]: What's the action item if this happens?

File: Library/Homebrew/utils/github.rb [hunk 7]
```diff
@@ -438,6 +440,38 @@ def dispatch_event(user, repo, event, **payload)
...
+    artifact_url = artifacts["artifacts"].first["archive_download_url"]
```
Comment (MikeMcQuaid) [inline Library/Homebrew/utils/github.rb]: Could `artifacts["artifacts"]` be empty?

File: Library/Homebrew/utils/github.rb [hunk 8]
```diff
@@ -438,6 +440,38 @@ def dispatch_event(user, repo, event, **payload)
...
+    when :env_token
+      curl_args = { header: "Authorization: token #{token}" }
+    end
+
+    FileUtils.chdir dir do
+      curl_args[:cache] = Pathname.new(dir)
+      curl_args[:secrets] = [token]
+      downloader = CurlDownloadStrategy.new(artifact_url, "artifact", pr, **curl_args)
+      downloader.fetch
+      downloader.stage
```
Comment (MikeMcQuaid) [inline Library/Homebrew/utils/github.rb]: Can you explain the `stage` here? Not sure I understand why/the format of what's being downloaded.

Comment (Bo98) [reply Library/Homebrew/utils/github.rb]: The artefact is a single archive which itself contains the bottle archives and JSONs.

Comment (MikeMcQuaid) [reply Library/Homebrew/utils/github.rb]: Gotcha. In that case I think I'd probably prefer using `curl` and `system "tar"`; this way it's not quite obvious what's going on and why.

Comment (jonchang) [reply Library/Homebrew/utils/github.rb]: I've added some comments to clarify what's going on here. I would prefer to use `download_strategy` as this stores the artifact archive in the Homebrew cache, meaning that a second invocation of `brew pr-pull` won't re-download a (possibly huge) set of bottles.

File: Library/Homebrew/dev-cmd/pr-pull.rb (hunk 4 above)
Comment (jonchang) [reply Library/Homebrew/dev-cmd/pr-pull.rb]: I've added a new `bintray.rb` that implements the Bintray API. `brew test-bot` can later be updated to use these new functions.

File: Library/Homebrew/utils/github.rb (hunk 8 above)
Comment (MikeMcQuaid) [reply Library/Homebrew/utils/github.rb]: Ah, cool. Good enough reason for me 👍

Comment (MikeMcQuaid) [reply Library/Homebrew/utils/github.rb]: Oh, only issue with that: can you confirm it gets handled by `brew cleanup`?

File: Library/Homebrew/bintray.rb [hunk 9]
```diff
@@ -0,0 +1,111 @@
+# frozen_string_literal: true
+
+require "utils/curl"
+require "json"
+
+class Bintray
```
Comment (MikeMcQuaid) [inline Library/Homebrew/bintray.rb]: Very nice 👏. Would love some basic tests.

File: Library/Homebrew/dev-cmd/pr-pull.rb [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+
+require "cli/parser"
+require "utils/github"
+require "tmpdir"
+require "bintray"
+
+module Homebrew
+  module_function
+
+  def pr_pull_args
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/pr-pull.rb]: Could we get the usual command args parse tests?

File: Library/Homebrew/bintray.rb [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+  API_URL = "https://api.bintray.com"
+
+  def inspect
+    "#<Bintray: user=#{@bintray_user} org=#{@bintray_org} key=***>"
+  end
+
+  def initialize(user = nil, key = nil, org = nil, clear = true)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/bintray.rb]:
```suggestion
  def initialize(user: nil, key: nil, org: nil, clear: true)
```

File: Library/Homebrew/utils/github.rb (hunk 8 above)
Comment (jonchang) [reply Library/Homebrew/utils/github.rb]: Yeah, already tested it locally. It gets scrubbed by `cleanup_unreferenced_downloads`

**Comment by jonchang** [comment]
In this new iteration:

* more closely matches the command line interface of `brew pull` with new `--branch-okay`, `--resolve` and other options
* added tests
* more robust checks on the github API for scenarios where there are e.g., missing artifacts or some such

File: Library/Homebrew/bintray.rb [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+require "json"
+
+class Bintray
+  API_URL = "https://api.bintray.com"
+
+  def inspect
+    "#<Bintray: user=#{@bintray_user} org=#{@bintray_org} key=***>"
+  end
+
+  def initialize(user: nil, key: nil, org: nil, clear: true)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/bintray.rb]: Sorry to contradict myself but what about:
```suggestion
  def initialize(user: ENV["HOMEBREW_BINTRAY_USER"], key: ENV["HOMEBREW_BINTRAY_KEY"], org: "homebrew", clear: true)
```

**Comment by jonchang** [comment]
All library functions now `raise Error` instead of `odie`.

File: Library/Homebrew/bintray.rb [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+    @bintray_user = user || ENV["HOMEBREW_BINTRAY_USER"]
+    @bintray_key = key || ENV["HOMEBREW_BINTRAY_KEY"]
+
+    if !@bintray_user || !@bintray_key
+      raise Error, "Missing HOMEBREW_BINTRAY_USER or HOMEBREW_BINTRAY_KEY variables!" unless Homebrew.args.dry_run?
```
Comment (MikeMcQuaid) [inline Library/Homebrew/bintray.rb]:
```suggestion
      raise UsageError, "Missing HOMEBREW_BINTRAY_USER or HOMEBREW_BINTRAY_KEY variables!" unless Homebrew.args.dry_run?
```

**Comment by jonchang** [comment]
Thanks for the reviews. Will merge this when 🍏

**Comment by jonchang** [comment]
This is now merged! Thanks for the help!!

Note that this is the last Homebrew/brew change needed prior to decommissioning Jenkins. We could turn it off now but I'd like to have this bake for a week or two so maintainers can sniff out bugs.

**Comment by Bo98** [comment]
Is there still a plan for Homebrew (Catalina) Testing?

**Comment by jonchang** [comment]
Yeah, I think a new command or something that uses `dispatch_event` plus a new `keep_old` payload in the workflow or something like that. I haven't really worked out the details yet but if you're interested in working on it we can hash something out together!

**Comment by Bo98** [comment]
Sure thing. Myself and FX use it a fair bit.
